Checkinout table

allenmichael099 opened the pull request:
Made the a table tab to display  bike checkins as a serparate tab from participants and trips. 

From my commit:
- Added function query_checkinout to utils_query_databasee
- Added the checkinout field to data_r in mod_load_data
- called mod_DT_server using data = data_r$checkinout
-  added a tabPanel for "Bike Check In" to app_ui
- moved the assert_flag for anon_locations within run_app

allenmichael099 committed:
Bike check In Table

- Added function query_checkinout to utils_query_databasee
- Added the checkinout field to data_r in mod_load_data
- called mod_DT_server using data = data_r$checkinout
-  added a tabPanel for "Bike Check In" to app_ui
- moved the assert_flag for anon_locations within run_app

shankari commented on R/app_server.R:
```diff
@@ -126,10 +126,12 @@ app_server <- function( input, output, session ) {
                  dplyr::select(-dplyr::any_of(cols_to_remove_from_trips_table)) %>%
                  sf::st_drop_geometry())
+    callModule(mod_DT_server,"DT_ui_checkinout",
+               data = data_r$checkinout)
```
```
>>> for t in scfg["supplementary_tables"]:
...     print("mod_DT_server", "DT_ui_%s" % t["name"], "data = data_r$%s" % t['name'])
...
mod_DT_server DT_ui_Checkinout data = data_r$Checkinout
mod_DT_server DT_ui_PolarBear data = data_r$PolarBear
mod_DT_server DT_ui_SuggestionSys data = data_r$SuggestionSys
mod_DT_server DT_ui_Tier_Sys data = data_r$Tier_Sys
```

shankari commented on R/app_ui.R:
```diff
@@ -114,6 +114,12 @@ app_ui <- function(request) {
+                        status = "primary",
+                        title = "Bike Check In",
+                        value = "checkinout",
+                        mod_DT_ui("DT_ui_checkinout")
```
```
>>> supplementary_tabs = ["tabPanel(status='primary', title=%s, value=%s, mod_DT_ui('DT_ui_%s')" % (t["tab_name"], t["name"], t["name"]) for t in scfg["supplementary_tables"]]
>>> supplementary_tabs
["tabPanel(status='primary', title=Bike Check in, value=Checkinout, mod_DT_ui('DT_ui_Checkinout')", "tabPanel(status='primary', title=Polar Bear, value=PolarBear, mod_DT_ui('DT_ui_PolarBear')", "tabPanel(status='primary', title=Suggestions, value=SuggestionSys, mod_DT_ui('DT_ui_SuggestionSys')", "tabPanel(status='primary', title=Leaderboard, value=Tier_Sys, mod_DT_ui('DT_ui_Tier_Sys')"]
>>> ....expand(supplementary_tabs)
```

shankari commented on R/mod_load_data.R:
```diff
@@ -53,6 +53,10 @@ mod_load_data_server <- function(input, output, session, cons){
+    message("About to load checkinout")
+    data_r$checkinout <- query_checkinout(cons) %>% data.table::setcolorder(c("user_id"))
+    message("Finished loading checkinout")
+    
```
```
>>> for t in scfg["supplementary_tables"]:
...     print("About to load %s" % t["name"])
...     print("data_r$%s = " % (t["name"]), "query_supplementary(cons, t['name'])")
...     print("Finished loading %s" % t["name"])
...
About to load Checkinout
data_r$Checkinout =  query_supplementary(cons, t['name'])
Finished loading Checkinout
About to load PolarBear
data_r$PolarBear =  query_supplementary(cons, t['name'])
Finished loading PolarBear
About to load SuggestionSys
data_r$SuggestionSys =  query_supplementary(cons, t['name'])
Finished loading SuggestionSys
About to load Tier_Sys
data_r$Tier_Sys =  query_supplementary(cons, t['name'])
Finished loading Tier_Sys
```

shankari commented on R/utils_query_database.R:
```diff
@@ -38,6 +38,15 @@ query_cleaned_trips <- function(cons) {
 
+#' @rdname query
+#' @export
+query_checkinout <- function(cons){
+  # cons <- connect_stage_collections(url = getOption('emdash.mongo_url')) 
+  cons$Checkinout$find('{}') %>%
+    as.data.table() %>%
+    normalise_uuid()
+}
+
```
```
query_supplementary <- function(cons, name) {
     cons[name].find('{}')....
}
```

shankari reviewed (commented):
this review is thinking through what the code changes would look like if the supplementary tables were in the config. I'm working with the config below and using python code.

```
supplementary_tables:
    - Checkinout:
        tab_name: "Bike Check in"
    - PolarBear:
        tab_name: "Polar Bear"
    - SuggestionSys:
        tab_name: "Suggestions"
    - Tier_Sys:
        tab_name : "Leaderboard"
```

shankari commented on R/utils_connect_database.R:
```diff
@@ -41,11 +41,17 @@ create_connection <- function(db, collection, url) {
 
 COLLECTIONS <- 
-  c("Stage_Profiles",
+  c("CarbonUsage",
+    "Checkinout",
+    "New_tier",
+    "PolarBear",
```
These are all specific to the gamified version that is used in CanBikeCo and should not be in the standard dashboard. Focusing on the supplemental tables,

```
>>> COLLECTIONS=["Stage_timeseries", "Stage_analysis_timeseries", "Stage_profile"]
>>> COLLECTIONS.extend([t["name"] for t in scfg["supplementary_tables"]])
>>> COLLECTIONS
['Stage_timeseries', 'Stage_analysis_timeseries', 'Stage_profile', 'Checkinout', 'PolarBear', 'SuggestionSys', 'Tier_Sys']
```

shankari reviewed (commented):
One more!

shankari commented:
Seems like it should all work. Hopefully R has a way to retrieve keys by name (e.g. the equivalent of `data[t["name"]]`). I'm not sure if the `data_r$t["name"]` would work 😄

asiripanich commented:
> 
> 
> Seems like it should all work. Hopefully R has a way to retrieve keys by name (e.g. the equivalent of `data[t["name"]]`). I'm not sure if the `data_r$t["name"]` would work 😄

Yes, that works in R as well `data_r$t[["name"]]`.

asiripanich commented on R/app_server.R:
```diff
@@ -80,11 +80,18 @@ app_server <- function( input, output, session ) {
+  # Convert the config's label map to a named vector. 
+  # The names are the original labels, and the vector values are the new labels.
+  named_label_vector <- unlist(getOption('emdash.col_labels_for_participts'))
+  originalColumnNames <- names(named_label_vector)
```
Let's stick with snake_case for variable naming.

allenmichael099 committed:
Dynamically adding tables

attempted to use renderUI and uiOutput

asiripanich commented:
@allenmichael099 Once I merge #32 to master, could you please pull new commits from master to this branch? Thanks

allenmichael099 committed:
Programmatic Table Addition

app_server:
- looped over each table
- used appendTab within app_server to add each new table, then called mod_DT_server for it
- moved normalize uuid and setcolorder from query_supplementary to the line after query supplementary was called within mod_load_data

allenmichael099 committed:
Moved config to inst

allenmichael099 committed:
Merge branch 'master' of https://github.com/asiripanich/emdash into Checkinout_Table

allenmichael099 commented:
After merging, I ran into an issue related to tidy_cleaned_trips and am currently looking into it. My first error was:
Warning: Error in setorderv: some columns are not in the data.table: data.end_fmt_time

allenmichael099 commented:
So the difference is within query_cleaned_trips: my working version said analysis/cleaned_trip, but after I merged it said analysis confirmed_trip, which returned a null data frame %>% null data table %>% couldn't use setorder. I'm not sure whether or not this has to do with differences between test-data and the actual data set.

There also is an issue just before generate trajectories so that when generate_trajectories is called, tidied_trips is missing 'fmt_time_utc', which gives:
 Warning: Error in eval: object 'fmt_time_utc' not found

shankari commented:
@allenmichael099 aha! As part of https://github.com/asiripanich/emdash/pull/23
I switched to reading `confirmed_trip` instead of `cleaned_trip`
I also updated the test database
<img width="1209" alt="Screen Shot 2021-06-08 at 4 03 34 PM" src="https://user-images.githubusercontent.com/2423263/121268578-17bd2c80-c873-11eb-85bc-586f2293bf34.png">

You need to update your test mongodb and then everything should work 👍 
This will also display `mode_confirm`, `purpose_confirm` etc in the trip table

allenmichael099 committed:
Revert "Merge branch 'master' of https://github.com/asiripanich/emdash into Checkinout_Table"

This reverts commit 3f9cbb1062dd052e233e40e0a871a9cacca8f690, reversing
changes made to 36636e522f3e165f2e200c19b55ee3adf1250692.

allenmichael099 committed:
Revert "Revert "Merge branch 'master' of https://github.com/asiripanich/emdash into Checkinout_Table""

This reverts commit 15c7fa52e2e0547767454edbe86b2251e8c2f48e.

allenmichael099 committed:
Small Fixes

- Removed Suggestion and Tier Sys from the default config and commented them out of app_config.yml

-  mod_load_data: Removed extra load participants and the extra generate_trajectories comment

- utils_query_database: Changed confirmed_trip to cleaned_trip within query_cleaned_trips

allenmichael099 committed:
cleaned_trip -> confirmed_trip

- within query_cleaned_trips, changed cleaned_trip to confirmed_trip
- commented out polar bear in the default yaml config

allenmichael099 commented:
@shankari Got it, thanks! I changed cleaned trip back to confirmed trip and downloaded the new test data. Everything looks like it works now

asiripanich commented on inst/app_config.yml:
```diff
@@ -0,0 +1,85 @@
+# Use this file to set options for the app
```
@allenmichael099 there is already a config-default.yml in inst. Maybe we should merge the two?

asiripanich replied:
Or is there any particular reason why you wanted to put this config file?

allenmichael099 replied:
No particular reason, I just wanted to keep the last two table suggestions somewhere. I'll take it out now.

asiripanich reviewed (commented):
Tested! I just have a few comments, but overall the PR looks good to me.

asiripanich commented on R/app_server.R:
```diff
@@ -125,15 +125,45 @@ app_server <- function(input, output, session) {
+
+    # For each supplementary table, append a new tabPanel and run the server function
+    # that specifies table behavior
+    for (t in tableList){
```
Now that we can have supplementary tables, you may want to look into how we can keep the plot panel above in sync with the table below.

allenmichael099 committed:
Delete app_config.yml

allenmichael099 committed:
Update config-default.yml

-removed an extra quote

shankari commented:
@asiripanich I am going to merge this into a local branch for deployment now

shankari committed:
Merge branch 'master' into Checkinout_Table

shankari commented:
This works great barring the sync between the table and the plot. I even enabled the commented out "Polar Bear" table and it worked just fine.

shankari commented:
Looked at the table with the end user today, and everything is fine except that the timestamp is in milliseconds.
We should really handle that in the dashboard instead of forcing the user to convert manually.

There are two ways in which we can do this:
- convert on the client (perhaps automatically by looking at fields called `ts`)
- store the formatted time on the server

In both cases, we need to decide how to specify the timezone for the conversion

One option is to have the phone send up the formatted time and the timezone, similar to the current background sensed entries. This would require no further customization on the dashboard.

shankari commented:
Timezones are tricky to work with.
The classic solution for time display is that you store everything as a `ts` and convert at the presentation layer, before displaying to the user.
BUT when you display to the user, **which timezone should you use?**

- no timezone = UTC = time in GMT. This is what AWS etc do, for example. This will be the same no matter where in the world you view it from, which is great for consistency, but requires the user to do mental math to figure out the conversion to their local timezone.
- timezone of the viewer. you can determine this from the computer/browser settings and just convert. This is what most TZ aware code (e.g. Github) does.
- timezone of the checkin/out. So if the user checked out in MDT, show this in MDT even if the viewer is in PDT. That's what we do with the trip start/end. If we want to do this, the fmt time and/or the timezone have to be generated from the app and stored along with the timestamp.

shankari commented:
This is only really applicable to app-generated supplementary data. Not sure how common a use case that is, and how people are going to want to use it in the future. Let's not overengineer. If @allenmichael099 can get the formatted time in the timezone of the computer along with automatically converting all `ts` columns, that is good enough for now.

If we need to send `fmt_time` from the app, we can migrate the server side data later.
